{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "U3LndOOk717h"
      },
      "source": [
        "\n",
        "# **MINI PROJECT in bioinformatics: part 1 (with changes for part 2)**\n",
        "\n",
        "## Data processing:\n",
        "\n",
        "First, initialize the data structures below:\n",
        "\n",
        "**original_data_dict**: dictionary that stores the lines from the data files\n",
        "\n",
        "- key= genome name\n",
        "\n",
        "- value= list of words of the current genome (each word is a list of cogs)\n",
        "\n",
        "\n",
        "**gen_uid_dict:**\n",
        "- key = genome name\n",
        "- value = uid\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "dGtlkf3RmAan"
      },
      "outputs": [],
      "source": [
        "import sys\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uoxF-isFC9vX"
      },
      "source": [
        "**process_data function**:\n",
        "\n",
        "-for each line in the file we insert to **original_data_dict** by the genome_NC.\n",
        "\n",
        "-insert the genome_NC and uid into **gen_uid_dict**\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "l73Hob7qmF5V"
      },
      "outputs": [],
      "source": [
        "def process_data(file, gen_uid_dict, original_data_dict):\n",
        "    with open(file) as my_file:\n",
        "        for line in my_file:\n",
        "            splited = line.split(\"#\")\n",
        "            genome_number = splited[1]\n",
        "            line_of_uid = splited[-2]\n",
        "            splited_uid_line = line_of_uid.split(\"_\")\n",
        "            my_uid = splited_uid_line[-1]\n",
        "            word_segment = splited[-1]\n",
        "            gen_uid_dict[genome_number] = my_uid\n",
        "            splited_word = word_segment.split(\"\\t\")\n",
        "            without_first = splited_word[1:-1]\n",
        "\n",
        "            if genome_number in original_data_dict:\n",
        "                original_data_dict[genome_number].append(without_first)\n",
        "            else:\n",
        "                my_list = [without_first]\n",
        "                original_data_dict[genome_number] = my_list\n",
        "    my_file.close\n",
        "\n",
        "orig_stdout = sys.stdout\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4PTjOPfTFWtZ"
      },
      "source": [
        "##**find clusters(s,d,q) function:**\n",
        "\n",
        "Given a set of COG-spelled genomes S, where each\n",
        "genome is segmented into segments such that each segment could contain one or\n",
        "more operons of the genomes in the database\n",
        "Find all gene\n",
        "clusters ð‘¤ (over the COG alphabet) of length ð‘‘ that are conserved in â‰¥ ð‘ž of the\n",
        "genomes in the database ð‘†.\n",
        "\n",
        "S = original_data_dict\n",
        "**all_clusters_dict:**\n",
        "\n",
        "- key = cluster (set of cogs)\n",
        "\n",
        "- value: an **inner dicitionary**: key = order of cogs,\n",
        "value = list of genomes numbers\n",
        "\n",
        "\n",
        "**final_clusters_dict:** (clusters that appears at least q times)\n",
        "\n",
        "- key = cluster (set of cogs)\n",
        "\n",
        "- value: an **inner dicitionary**: key = order of cogs,\n",
        "value = list of genomes numbers\n",
        "\n",
        "\n",
        "\n",
        "**genome_substrings_dict:** (for each genome we store clusters and the first order we found)\n",
        "\n",
        "- key = cluster: set of cogs\n",
        "- value = order of cogs\n",
        "\n",
        "\n",
        "**clusters_counters_dict:** (number of appearnces in all genomes for each cluster)\n",
        "\n",
        "- key = cluster: set of cogs\n",
        "\n",
        "- value = number of appearnces\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "q5OWH-J4mOZY"
      },
      "outputs": [],
      "source": [
        "def find_clusters(s, d, q, final_clusters_dict, all_clusters_dict, clusters_counter_dict):\n",
        "    for curr_genome in s:\n",
        "        genome_substrings_dict = {}\n",
        "        for segment in s[curr_genome]: #iterate over different lines of the same genome\n",
        "            for i in range(len(segment) - d+1):#sliding window\n",
        "                curr_d_mer = segment[i:i + d] #create the current dmer\n",
        "                temp_dict = {}\n",
        "                my_str_order = \"\"\n",
        "                if 'X' not in curr_d_mer and len(curr_d_mer)==d: #ignore dmers that contains X\n",
        "                    cluster_set = set()\n",
        "                    for cog in curr_d_mer:# add each cog to set represent the cluster\n",
        "                        my_str_order += cog\n",
        "                        if cog != '':\n",
        "                            cluster_set.add(cog)\n",
        "                    frozen_cluster = frozenset(cluster_set)\n",
        "                    if len(frozen_cluster) == d:#make sure the set in size d (can be cases cog is repeated and size of set is smaller)\n",
        "                        if frozen_cluster not in genome_substrings_dict: #found a new cluster\n",
        "                            genome_substrings_dict[frozen_cluster] = my_str_order\n",
        "\n",
        "        for cluster in genome_substrings_dict:\n",
        "            my_hashmap = {}\n",
        "            curr_order = genome_substrings_dict[cluster]\n",
        "            #maintain the number of appearance for each cluster\n",
        "            if cluster in clusters_counter_dict:\n",
        "                clusters_counter_dict[cluster] += 1\n",
        "            else:\n",
        "                clusters_counter_dict[cluster] = 1\n",
        "            #add cluster to clusters dictionary and then add genome number to the list of the cluster by order\n",
        "            if cluster not in all_clusters_dict:\n",
        "                my_hashmap[curr_order] = [curr_genome]\n",
        "                all_clusters_dict[cluster] = my_hashmap\n",
        "            else: #cluster in dict\n",
        "                curr_cluster = all_clusters_dict[cluster]\n",
        "                if curr_order in curr_cluster:\n",
        "                    curr_cluster[curr_order].append(curr_genome)\n",
        "                else:\n",
        "                    curr_cluster[curr_order] = [curr_genome]\n",
        "    #add to final_clusters_dict clusters that appear at least q times\n",
        "\n",
        "    #change clusters_counter_dict to include only clusters that apears at least q times\n",
        "    clusters_counter_dict = {cluster: value for cluster, value in clusters_counter_dict.items() if value >= q}\n",
        "    sorted_list = sorted(clusters_counter_dict, key=lambda key: clusters_counter_dict[key], reverse=True)\n",
        "    sys.stdout = orig_stdout\n",
        "\n",
        "    return sorted_list"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bI6A1MzQbwdI"
      },
      "source": [
        "##**profile by gene cluster:**\n",
        "subroutine that, given a gene cluster, computes and prints a descriptive â€œprofileâ€ of the cluster. For each cluster we report, the output include a â€œcharacterization profileâ€ including the following information:\n",
        "_____________________________________________________________________\n",
        "1) The rank of the gene cluster in the sorted list of all gene clusters found for the given parameters (d,q) and the input genome file and the number of instances it has among the genomes in the input file.\n",
        "\n",
        "**create_clusters_rank_dict function:**\n",
        "\n",
        "create the **rank_dict** that contain : key = cluster, value = rank of the cluster.\n",
        "______________________________________________________________________\n",
        "2) Pattern information: Print the set of COGs participating in the pattern of the gene cluster\n",
        "\n",
        "**create_cog_dict function:**\n",
        "\n",
        "create the **cog_dic** that contain: key = cog, value = line from \"COG_INFO_TABLE\"\n",
        "\n",
        "**get_pattern_info(cluster):**\n",
        "\n",
        "for each cog in cluster set, prints the coresponding information from **cog_dict**\n",
        "\n",
        "_______________________________________________________________________\n",
        "3) Report the various gene orders of the pattern in the instances and the percentage of instances found per each gene order\n",
        "\n",
        " **get_various_gene_orders(cluster):**\n",
        "\n",
        "get the gene orders from **final_clusters_dict[cluster]** and calculate the percentage.\n",
        "\n",
        " For each of the distinct orders in which the gene cluster was found, report distribution statistics of its abundance in phyla and habitats:\n",
        "\n",
        " A. Report the various phyla and the percentage of instances per this order found in each phylum.\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        " B. Report the environmental descriptions of the instances of the gene cluster in which the order is found and the percentage of instances of the gene cluster in which the order is found per each environmental category.\n",
        "\n",
        "**create_taxa_dict function:**\n",
        "\n",
        "create the **taxa_dict**: key = uid, value = phylum\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "**create_habitat_dict function:**\n",
        "\n",
        "create the **habitat_dict** : key = uid, value = phylum\n",
        "\n",
        "**get_taxa(cluster):**\n",
        "- for each order found for the cluster, we iterate over genes, get the phyla using the **taxa_dict** by gene uid (gen_uid_dict[genome]).\n",
        "\n",
        "- calculate the percent of each phylum, by maintaining a dictionary called **phylas**, that counts the amount of the different phylums.\n",
        "\n",
        "\n",
        "**get_habitat(cluster):**\n",
        "- for each order found for the cluster, we iterate over genes, get the habitat using the **habitat_dict** by gene uid (gen_uid_dict[genome]).\n",
        "\n",
        "- calculate the percent of each habitat, by maintaining a dictionary called **habitats**, that counts the amount of the different habitats.\n",
        "\n",
        "\n",
        "##**get_cluster_profile function:**\n",
        "\n",
        "for the given cluster, all the requested information is printed, by calling the **get** functions above"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "sG-zMyd9nk9U"
      },
      "outputs": [],
      "source": [
        "\n",
        "def create_clusters_rank_dict(clusters_counter_dict, rank_dict): # C1\n",
        "    sorted_list = sorted(clusters_counter_dict, key=lambda key: clusters_counter_dict[key], reverse= True)\n",
        "    for i in range(len(sorted_list)):\n",
        "        rank_dict[sorted_list[i]] = i+1\n",
        "    return rank_dict\n",
        "\n",
        "\n",
        "def create_cog_dict(cog_dict):  # C2\n",
        "    with open(\"COG_INFO_TABLE.txt\",'r', encoding ='latin-1') as cog_file:\n",
        "        data_cogs_lines = cog_file.readlines()\n",
        "        for cog_line in data_cogs_lines:\n",
        "            splited_line = cog_line.split(\";\")\n",
        "            cog_number = splited_line[0]\n",
        "            cog_dict[cog_number] = cog_line\n",
        "    cog_file.close\n",
        "\n",
        "def get_pattern_info(cluster,cog_dict):\n",
        "    sorted_cluster = sorted(map(int, set(cluster)))#sort cogs\n",
        "    for cog in sorted_cluster:\n",
        "        my_cog_str = str(cog)\n",
        "        #add zeros to start of cog if needed\n",
        "        to_add = 4-len(str(cog))\n",
        "        for i in range(to_add):\n",
        "            my_cog_str = '0'+ my_cog_str\n",
        "        key = \"COG\" + my_cog_str\n",
        "        print(my_cog_str, cog_dict[key], end=\"\")\n",
        "\n",
        "\n",
        "def create_taxa_dict(taxa_dict):  # C3A\n",
        "    with open(\"taxa.txt\") as taxa_file:\n",
        "        lines = taxa_file.readlines()\n",
        "        for i in range(1, len(lines)):\n",
        "            splited = lines[i].split(\",\")\n",
        "            uid_line = splited[-4]\n",
        "            phylum = splited[1]\n",
        "            uid_word = uid_line.split(\"_\")\n",
        "            uid = uid_word[-1]\n",
        "            taxa_dict[uid] = phylum\n",
        "    taxa_file.close\n",
        "\n",
        "\n",
        "def create_habitat_dict(habitat_dict):  # C3A\n",
        "    with open(\"habitat.txt\") as habitat_file:\n",
        "        lines = habitat_file.readlines()\n",
        "        for i in range(1, len(lines)):\n",
        "            splited = lines[i].split(\";\")\n",
        "            my_habitat = splited[-1]\n",
        "            uid_line = splited[1]\n",
        "            splited_uid = uid_line.split(\"_\")\n",
        "            uid = splited_uid[-1]\n",
        "            habitat_dict[uid] = my_habitat\n",
        "    habitat_file.close\n",
        "\n",
        "\n",
        "def get_taxa(cluster, final_clusters_dict, taxa_dict, gen_uid_dict):\n",
        "    cluster_dict = final_clusters_dict[cluster]\n",
        "    for order in cluster_dict:\n",
        "        phylas = {}\n",
        "        for genome in cluster_dict[order]:\n",
        "            my_uid = gen_uid_dict[genome]\n",
        "            my_phyla = taxa_dict[my_uid]\n",
        "            if my_phyla in phylas:\n",
        "                phylas[my_phyla] +=1\n",
        "            else:\n",
        "                phylas[my_phyla] = 1\n",
        "        total = len(cluster_dict[order])\n",
        "        print(\"order: \", order)\n",
        "        for phyla in phylas:\n",
        "            percent = round(phylas[phyla] / total * 100, 2)\n",
        "            print(\"phylum: \", phyla, end=\" \")\n",
        "            print(\"percentage: \", percent, \"%\")\n",
        "\n",
        "\n",
        "def get_habitat(cluster, final_clusters_dict, habitat_dict, gen_uid_dict):\n",
        "    cluster_dict = final_clusters_dict[cluster]\n",
        "    for order in cluster_dict:\n",
        "        habitats = {}\n",
        "        for genome in cluster_dict[order]:\n",
        "            my_habitat = habitat_dict[gen_uid_dict[genome]]\n",
        "            if my_habitat in habitats:\n",
        "                habitats[my_habitat] +=1\n",
        "            else:\n",
        "                habitats[my_habitat] = 1\n",
        "        total = len(cluster_dict[order])\n",
        "        print(\"order: \", order)\n",
        "        for habitat in habitats:\n",
        "            percent = round(habitats[habitat] / total * 100, 2)\n",
        "            print(\"environment: \", habitat, end=\"\")\n",
        "            print(\"percentage: \", percent, \"%\")\n",
        "\n",
        "\n",
        "def get_various_gene_orders(cluster,final_clusters_dict, clusters_counter_dict):\n",
        "    my_dict = final_clusters_dict[cluster]\n",
        "    total = clusters_counter_dict[cluster]\n",
        "    for order in my_dict:\n",
        "        percent = round(len(my_dict[order]) / total * 100, 2)\n",
        "        print(order, \"percentage of order: \", percent, \"%\")\n",
        "\n",
        "\n",
        "def get_cluster_profile(cluster, rank_dict, clusters_counter_dict, final_clusters_dict, gen_uid_dict, taxa_dict, habitat_dict,cog_dict):\n",
        "    print(\"Cluster: \")\n",
        "    #sort set of cluster by COG id\n",
        "    sorted_cluster = sorted(map(int, set(cluster)))\n",
        "    cogs_list = []\n",
        "    for cog in sorted_cluster:\n",
        "        my_cog_str = str(cog)\n",
        "        to_add = 4-len(str(cog))\n",
        "        for i in range(to_add):\n",
        "            my_cog_str = '0'+ my_cog_str\n",
        "        cogs_list.append(my_cog_str)\n",
        "    print(cogs_list)\n",
        "    print(\"Rank: \", rank_dict[cluster])\n",
        "    print(\"Appearance amount:\", clusters_counter_dict[cluster])\n",
        "    print(\"\")\n",
        "    print(\"Pattern info: \")\n",
        "    get_pattern_info(cluster,cog_dict)\n",
        "    print(\"\")\n",
        "    print(\"Instance information: \")\n",
        "    print(\"\")\n",
        "    print(\"Taxonomical info: \")\n",
        "    get_taxa(cluster, final_clusters_dict, taxa_dict, gen_uid_dict)\n",
        "    print(\"\")\n",
        "    print(\"Various gene_order: \")\n",
        "    get_various_gene_orders(cluster, final_clusters_dict, clusters_counter_dict)\n",
        "    print(\"\")\n",
        "    print(\"environmental description: \")\n",
        "    get_habitat(cluster, final_clusters_dict, habitat_dict, gen_uid_dict)\n",
        "    print(\"\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "C6oP-U8gsKTu"
      },
      "source": [
        "##**Run the  program**\n",
        "\n",
        " extract all gene clusters of cardinality ranging from d=2 up to\n",
        "d=4, using q=5 from the file of plasmid genomes, and for each cardinality value,\n",
        "report the 4 most abundant gene clusters found, sorted by decreasing abundance\n",
        "order per each value of d.\n",
        "Print the reported gene clusters into an external output file, producing a separate ourtput for each cardinality value d, sorted in decreasing order by the number of instances per each gene cluster.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "pfTlU_pmqEBh"
      },
      "outputs": [],
      "source": [
        "def run_program(data_file):\n",
        "    original_data_dict = {}\n",
        "    all_clusters_dict = {}\n",
        "    final_clusters_dict = {}\n",
        "    gen_uid_dict = {}\n",
        "    clusters_counter_dict = {}\n",
        "    rank_dict = {}\n",
        "    cog_dict = {}\n",
        "    taxa_dict = {}\n",
        "    habitat_dict = {}\n",
        "\n",
        "    orig_stdout = sys.stdout\n",
        "    process_data(data_file, gen_uid_dict, original_data_dict)\n",
        "    create_cog_dict(cog_dict)\n",
        "    create_taxa_dict(taxa_dict)\n",
        "    create_habitat_dict(habitat_dict)\n",
        "\n",
        "\n",
        "    for d_val in range(2, 5):\n",
        "        all_clusters_dict = {}\n",
        "        final_clusters_dict = {}\n",
        "        clusters_counter_dict = {}\n",
        "        rank_dict = {}\n",
        "        find_clusters(original_data_dict, d_val, 5, final_clusters_dict, all_clusters_dict, clusters_counter_dict)\n",
        "        final_clusters_dict = {cluster: values for cluster, values in all_clusters_dict.items() if clusters_counter_dict[cluster] >= 5}\n",
        "        curr_rank_dict = create_clusters_rank_dict(clusters_counter_dict, rank_dict)#find rank of the diffrent clusters\n",
        "        keys = [key for key in curr_rank_dict]#make a list of clusters only from rank_dict\n",
        "        my_ranks = keys[:4] #The 4 most abundant clusters\n",
        "        file_name = \"gene_clusters_report_d=%s.txt\" % d_val\n",
        "        with open(file_name, 'w') as output_file:\n",
        "            sys.stdout = output_file# change the standart output to print into the new file\n",
        "            index = 1\n",
        "            for cluster in my_ranks:\n",
        "                get_cluster_profile(cluster, rank_dict, clusters_counter_dict, final_clusters_dict, gen_uid_dict, taxa_dict, habitat_dict,cog_dict)\n",
        "                index +=1\n",
        "            output_file.close()\n",
        "        sys.stdout = orig_stdout #return the standard output\n",
        "\n",
        "\n",
        "run_program(\"cog_words_plasmid.txt\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "eD7jg4ai3T6s"
      },
      "source": [
        "# **Part 2**\n",
        "**Data Preparation:** â€œCleanâ€ the file of bacterial genomes so that only genomes that have at\n",
        "least one plasmid in the data remain\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "rnTouwZ533Cn"
      },
      "outputs": [],
      "source": [
        "def clean_bac_genomes_file():\n",
        "    uid_from_plasmid = {}\n",
        "    with open(\"cog_words_plasmid.txt\", 'r') as plasmid_file:\n",
        "        for line in plasmid_file:\n",
        "            splited = line.split('#')\n",
        "            uid_line = splited[3]\n",
        "            splited_uid_line = uid_line.split('_')\n",
        "            #print(splited_uid_line)\n",
        "            uid = splited_uid_line[-1]\n",
        "            uid_from_plasmid.update({uid:\"\"})\n",
        "\n",
        "    with open('clean_file.txt', 'w') as clean_file:\n",
        "        with open(\"cog_words_bac.txt\",'r') as bac_file:\n",
        "            for line in bac_file:\n",
        "                splited = line.split('#')\n",
        "                uid_line = splited[3]\n",
        "                splited_uid_line = uid_line.split('_')\n",
        "                uid = splited_uid_line[-1]\n",
        "                if uid in uid_from_plasmid:\n",
        "                    clean_file.write(line)\n",
        "\n",
        "\n",
        "clean_bac_genomes_file()"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "5Zi32-cF1yJZ"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "3YPmqNal10Po"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pbQk-BwrfAMY"
      },
      "source": [
        "**our comparative analysis approach for gene cluster profiles:**\n",
        "\n",
        "We Propose a novel approach to compare two gene cluster profiles (we focus in case of the same cluster in 2 different  datasets) and produce a score (a number between 0 and 100) to measure how distinct the clusters are.\n",
        "the calculation is composed from the following parameters:\n",
        "\n",
        "\n",
        "1) **X (Genetic Variation Score):** This parameter evaluates the difference in abundance between various gene orders within the clusters. We utilize a formula that directly computes the absolute difference between the gene order percentages of the two profiles. For instance:\n",
        "\n",
        "    - Profile 1: [98.06%, 1.94%]\n",
        "    - Profile 2: [80%, 20%]\n",
        "\n",
        "   \n",
        "The calculation for **X** involves:\n",
        "\n",
        "(| 98.06 - 80 | + | 1.94 - 20 |) / 2\n",
        "\n",
        "2) **Y (Taxa Score):** This parameter assesses the dissimilarity in taxa parameters between the two datasets, while considering the percentages of different gene orders. Similar to **X**, we directly compute the absolute difference between the taxonomic composition percentages of the two profiles.\n",
        "\n",
        "3) **Z (Environment Score):** Here, we quantify the discrepancy in environmental parameters between the two datasets, while also factoring in the percentages of different gene orders. Like **X** and **Y**, the calculation involves directly computing the absolute difference between the environment percentages of the two profiles.\n",
        "\n",
        "The **final_score** is computed as the average of the three individual scores(X,Y,Z)\n",
        "\n",
        "\n",
        "This approach provides a straightforward assessment of the dissimilarity between gene cluster profiles, focusing solely on the absolute difference. The final score facilitates comparative analysis and aids in identifying clusters with notable distinctions across datasets.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "bmzgA4gDoBIp",
        "outputId": "73342267-911d-4d6f-8f6d-09be54b14c40"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "13\n",
            "34\n",
            "69.33333333333333\n",
            "The order:  1088120918981091\n",
            "Percentage of the order in the first set:  7.69\n",
            "Percentage of the order in the second set:  14.71\n",
            "The order:  1898108812091091\n",
            "Percentage of the order in the first set:  7.69\n",
            "Percentage of the order in the second set:  0\n",
            "The order:  1898108810911209\n",
            "Percentage of the order in the first set:  46.15\n",
            "Percentage of the order in the second set:  14.71\n",
            "The order:  1898109112091088\n",
            "Percentage of the order in the first set:  7.69\n",
            "Percentage of the order in the second set:  0\n",
            "The order:  1209189810911088\n",
            "Percentage of the order in the first set:  7.69\n",
            "Percentage of the order in the second set:  2.94\n",
            "The order:  1088109118981209\n",
            "Percentage of the order in the first set:  7.69\n",
            "Percentage of the order in the second set:  0\n",
            "The order:  1209189810881091\n",
            "Percentage of the order in the first set:  15.38\n",
            "Percentage of the order in the second set:  47.06\n",
            "The order:  1088109112091898\n",
            "Percentage of the order in the first set:  0\n",
            "Percentage of the order in the second set:  11.76\n",
            "The order:  1088120910911898\n",
            "Percentage of the order in the first set:  0\n",
            "Percentage of the order in the second set:  8.82\n",
            "Taxa_info:\n",
            "The order:  1088120918981091\n",
            "The phylum:  Proteobacteria\n",
            "Percentage of the phylum in the first set:  100.0\n",
            "Percentage of the phylum in the second set:  100.0\n",
            "The order:  1898108812091091\n",
            "The phylum:  Acidobacteria\n",
            "Percentage of the phylum in the first set:  100.0\n",
            "Percentage of the phylum in the second set:  0\n",
            "The order:  1898108810911209\n",
            "The phylum:  Proteobacteria\n",
            "Percentage of the phylum in the first set:  100.0\n",
            "Percentage of the phylum in the second set:  100.0\n",
            "The order:  1898109112091088\n",
            "The phylum:  Spirochaetes\n",
            "Percentage of the phylum in the first set:  100.0\n",
            "Percentage of the phylum in the second set:  0\n",
            "The order:  1209189810911088\n",
            "The phylum:  Fusobacteria\n",
            "Percentage of the phylum in the first set:  100.0\n",
            "Percentage of the phylum in the second set:  0\n",
            "The phylum:  Bacteroidetes\n",
            "Percentage of the phylum in the first set:  0\n",
            "Percentage of the phylum in the second set:  100.0\n",
            "The order:  1088109118981209\n",
            "The phylum:  Firmicutes\n",
            "Percentage of the phylum in the first set:  100.0\n",
            "Percentage of the phylum in the second set:  0\n",
            "The order:  1209189810881091\n",
            "The phylum:  Proteobacteria\n",
            "Percentage of the phylum in the first set:  100.0\n",
            "Percentage of the phylum in the second set:  0\n",
            "The phylum:  Firmicutes\n",
            "Percentage of the phylum in the first set:  0\n",
            "Percentage of the phylum in the second set:  100.0\n",
            "The order:  1088109112091898\n",
            "The phylum:  Proteobacteria\n",
            "Percentage of the phylum in the first set:  0\n",
            "Percentage of the phylum in the second set:  100.0\n",
            "The order:  1088120910911898\n",
            "The phylum:  Proteobacteria\n",
            "Percentage of the phylum in the first set:  0\n",
            "Percentage of the phylum in the second set:  100.0\n",
            "Env_info:\n",
            "The order:  1088120918981091\n",
            "The habitat:  Not Annotated\n",
            "\n",
            "Percentage of the habitat in the first set:  100.0\n",
            "Percentage of the habitat in the second set:  40.0\n",
            "The habitat:  Marine Environment\n",
            "\n",
            "Percentage of the habitat in the first set:  0\n",
            "Percentage of the habitat in the second set:  20.0\n",
            "The habitat:  Fresh water\n",
            "\n",
            "Percentage of the habitat in the first set:  0\n",
            "Percentage of the habitat in the second set:  20.0\n",
            "The habitat:  Soil and Sediment\n",
            "\n",
            "Percentage of the habitat in the first set:  0\n",
            "Percentage of the habitat in the second set:  20.0\n",
            "The order:  1898108812091091\n",
            "The habitat:  Not Annotated\n",
            "\n",
            "Percentage of the habitat in the first set:  100.0\n",
            "Percentage of the habitat in the second set:  0\n",
            "The order:  1898108810911209\n",
            "The habitat:  Plant\n",
            "\n",
            "Percentage of the habitat in the first set:  16.67\n",
            "Percentage of the habitat in the second set:  80.0\n",
            "The habitat:  Not Annotated\n",
            "\n",
            "Percentage of the habitat in the first set:  33.33\n",
            "Percentage of the habitat in the second set:  0\n",
            "The habitat:  Soil and Sediment\n",
            "\n",
            "Percentage of the habitat in the first set:  16.67\n",
            "Percentage of the habitat in the second set:  0\n",
            "The habitat:  Animal\n",
            "\n",
            "Percentage of the habitat in the first set:  16.67\n",
            "Percentage of the habitat in the second set:  0\n",
            "The habitat:  Marine Environment\n",
            "\n",
            "Percentage of the habitat in the first set:  16.67\n",
            "Percentage of the habitat in the second set:  0\n",
            "The habitat:  Fresh water\n",
            "\n",
            "Percentage of the habitat in the first set:  0\n",
            "Percentage of the habitat in the second set:  20.0\n",
            "The order:  1898109112091088\n",
            "The habitat:  Animal\n",
            "\n",
            "Percentage of the habitat in the first set:  100.0\n",
            "Percentage of the habitat in the second set:  0\n",
            "The order:  1209189810911088\n",
            "The habitat:  Soil and Sediment\n",
            "\n",
            "Percentage of the habitat in the first set:  100.0\n",
            "Percentage of the habitat in the second set:  0\n",
            "The habitat:  Human\n",
            "\n",
            "Percentage of the habitat in the first set:  0\n",
            "Percentage of the habitat in the second set:  100.0\n",
            "The order:  1088109118981209\n",
            "The habitat:  Host\n",
            "\n",
            "Percentage of the habitat in the first set:  100.0\n",
            "Percentage of the habitat in the second set:  0\n",
            "The order:  1209189810881091\n",
            "The habitat:  Soil and Sediment\n",
            "\n",
            "Percentage of the habitat in the first set:  50.0\n",
            "Percentage of the habitat in the second set:  25.0\n",
            "The habitat:  Host\n",
            "\n",
            "Percentage of the habitat in the first set:  50.0\n",
            "Percentage of the habitat in the second set:  43.75\n",
            "The habitat:  Human\n",
            "\n",
            "Percentage of the habitat in the first set:  0\n",
            "Percentage of the habitat in the second set:  18.75\n",
            "The habitat:  Plant\n",
            "\n",
            "Percentage of the habitat in the first set:  0\n",
            "Percentage of the habitat in the second set:  6.25\n",
            "The habitat:  Animal\n",
            "\n",
            "Percentage of the habitat in the first set:  0\n",
            "Percentage of the habitat in the second set:  6.25\n",
            "The order:  1088109112091898\n",
            "The habitat:  Soil and Sediment\n",
            "\n",
            "Percentage of the habitat in the first set:  0\n",
            "Percentage of the habitat in the second set:  25.0\n",
            "The habitat:  Animal\n",
            "\n",
            "Percentage of the habitat in the first set:  0\n",
            "Percentage of the habitat in the second set:  25.0\n",
            "The habitat:  Not Annotated\n",
            "\n",
            "Percentage of the habitat in the first set:  0\n",
            "Percentage of the habitat in the second set:  25.0\n",
            "The habitat:  Wastewater\n",
            "\n",
            "Percentage of the habitat in the first set:  0\n",
            "Percentage of the habitat in the second set:  25.0\n",
            "The order:  1088120910911898\n",
            "The habitat:  Animal\n",
            "\n",
            "Percentage of the habitat in the first set:  0\n",
            "Percentage of the habitat in the second set:  33.33\n",
            "The habitat:  Human\n",
            "\n",
            "Percentage of the habitat in the first set:  0\n",
            "Percentage of the habitat in the second set:  33.33\n",
            "The habitat:  Fresh water\n",
            "\n",
            "Percentage of the habitat in the first set:  0\n",
            "Percentage of the habitat in the second set:  33.33\n"
          ]
        }
      ],
      "source": [
        "import math\n",
        "various_gene_orders = {}\n",
        "taxa_info = {}\n",
        "env_info = {}\n",
        "def calculate_profile_score(cluster, q1,q2):\n",
        "    global various_gene_orders, taxa_info, env_info\n",
        "\n",
        "    num_of_orders = 0\n",
        "    max = 0\n",
        "    x = 0\n",
        "\n",
        "    for order in various_gene_orders:\n",
        "        x += math.sqrt((various_gene_orders[order][0]-various_gene_orders[order][1]) ** 2)\n",
        "\n",
        "    x = round(x / 2)\n",
        "\n",
        "    y = 0\n",
        "\n",
        "    for order in taxa_info :\n",
        "        i = 0\n",
        "        data = taxa_info[order]\n",
        "\n",
        "        for item in data:\n",
        "            my_list = data[item]\n",
        "            y += math.sqrt(((my_list[0]*various_gene_orders[order][0]/100 )-(my_list[1]*various_gene_orders[order][1]/100))** 2)\n",
        "\n",
        "\n",
        "    y = round(y / 2)\n",
        "\n",
        "    #print(y)\n",
        "\n",
        "    z = 0\n",
        "\n",
        "    for order in env_info:\n",
        "        i = 0\n",
        "        data = env_info[order]\n",
        "\n",
        "        for item in data:\n",
        "            my_list = data[item]\n",
        "            z += math.sqrt((my_list[0]*various_gene_orders[order][0]/100 - my_list[1]*various_gene_orders[order][1]/100) ** 2)\n",
        "\n",
        "\n",
        "    z = round(z / 2)\n",
        "\n",
        "    if x > 100 or y > 100 or z > 100:\n",
        "        print(\"x: \", x,\"y: \",y,\"z: \", z)\n",
        "    score = (x + y + z) / 3\n",
        "    return score\n",
        "\n",
        "\n",
        "def create_cluster_profile(s, cluster, taxa_info, env_info, various_gene_orders,index,q):\n",
        "    gen_uid_dict = {}\n",
        "    original_data_dict = {}\n",
        "    cog_dict = {}\n",
        "    taxa_dict = {}\n",
        "    habitat_dict = {}\n",
        "    final_clusters_dict = {}\n",
        "    all_clusters_dict = {}\n",
        "    clusters_counter_dict = {}\n",
        "    process_data(s, gen_uid_dict, original_data_dict)\n",
        "    create_cog_dict(cog_dict)\n",
        "    create_taxa_dict(taxa_dict)\n",
        "    create_habitat_dict(habitat_dict)\n",
        "    d_val = len(cluster)\n",
        "    find_clusters(original_data_dict, d_val, 1, final_clusters_dict, all_clusters_dict, clusters_counter_dict) # q=2 is a default value\n",
        "    final_clusters_dict = {cluster: values for cluster, values in all_clusters_dict.items() if clusters_counter_dict[cluster] >= q}\n",
        "    get_gene_orders_dict(cluster,index, final_clusters_dict, clusters_counter_dict)\n",
        "    get_taxa_info(cluster,index, final_clusters_dict, taxa_dict, gen_uid_dict)\n",
        "    get_env_info(cluster,index, final_clusters_dict, gen_uid_dict, habitat_dict)\n",
        "    print(clusters_counter_dict[cluster])\n",
        "\n",
        "\n",
        "def get_gene_orders_dict(cluster,index, final_clusters_dict, clusters_counter_dict):\n",
        "    global various_gene_orders\n",
        "    my_dict = final_clusters_dict[cluster]\n",
        "    total = clusters_counter_dict[cluster]\n",
        "    for order in my_dict:\n",
        "        percent = round(len(my_dict[order]) / total * 100, 2)\n",
        "        if order not in various_gene_orders:\n",
        "            various_gene_orders[order] = [0,0]\n",
        "            various_gene_orders[order][index] = percent\n",
        "        else:\n",
        "            various_gene_orders[order][index] = percent\n",
        "\n",
        "\n",
        "def get_taxa_info(cluster, index, final_clusters_dict, taxa_dict, gen_uid_dict):\n",
        "    global taxa_info\n",
        "    cluster_dict = final_clusters_dict[cluster]\n",
        "    for order in cluster_dict:\n",
        "        phylas = {}\n",
        "        for genome in cluster_dict[order]:\n",
        "            my_uid = gen_uid_dict[genome]\n",
        "            my_phyla = taxa_dict[my_uid]\n",
        "            if my_phyla in phylas:\n",
        "                phylas[my_phyla] += 1\n",
        "            else:\n",
        "                phylas[my_phyla] = 1\n",
        "        total = len(cluster_dict[order])\n",
        "       #-----------------------------------\n",
        "        if order not in taxa_info:\n",
        "            taxa_info[order] = {}\n",
        "        curr_order_dict = taxa_info[order]\n",
        "        for phyla in phylas:\n",
        "            percent = round(phylas[phyla] / total * 100, 2)\n",
        "            if phyla not in curr_order_dict:\n",
        "                curr_order_dict[phyla] = [0,0]\n",
        "                curr_order_dict[phyla][index] = percent\n",
        "            else:\n",
        "                curr_order_dict[phyla][index] = percent\n",
        "\n",
        "\n",
        "def get_env_info(cluster,index, final_clusters_dict, gen_uid_dict, habitat_dict):\n",
        "    global env_info\n",
        "    cluster_dict = final_clusters_dict[cluster]\n",
        "    for order in cluster_dict:\n",
        "        habitats = {}\n",
        "        for genome in cluster_dict[order]:\n",
        "            my_uid = gen_uid_dict[genome]\n",
        "            my_habitat = habitat_dict[my_uid]\n",
        "            if my_habitat in habitats:\n",
        "                habitats[my_habitat] += 1\n",
        "            else:\n",
        "                habitats[my_habitat] = 1\n",
        "            total = len(cluster_dict[order])\n",
        "            if order not in env_info:\n",
        "                env_info[order] = {}\n",
        "            curr_order_dict = env_info[order]\n",
        "            for habitat in habitats:\n",
        "                percent = round(habitats[habitat] / total * 100, 2)\n",
        "                if habitat not in curr_order_dict:\n",
        "                    curr_order_dict[habitat] = [0,0]\n",
        "                    curr_order_dict[habitat][index] = percent\n",
        "                else:\n",
        "                    curr_order_dict[habitat][index] = percent\n",
        "\n",
        "\n",
        "def test_score():\n",
        "    global taxa_info\n",
        "    global env_info\n",
        "    global various_gene_orders\n",
        "    various_gene_orders = {}\n",
        "    taxa_info = {}\n",
        "    env_info = {}\n",
        "    cluster_set = frozenset({'1898', '1088', '1209', '1091'})\n",
        "    create_cluster_profile(\"cog_words_plasmid.txt\", cluster_set, taxa_info, env_info, various_gene_orders,0,1)\n",
        "    create_cluster_profile(\"clean_file.txt\", cluster_set, taxa_info, env_info, various_gene_orders,1,10)\n",
        "    print(calculate_profile_score(cluster_set,1,10))\n",
        "\n",
        "\n",
        "#test_score()\n",
        "#the target of this function is to check whether the function calculate_score() is working correctly"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BG7eqWM1X31Y"
      },
      "source": [
        "**3)** Based on your code from Part 1, write a program\\pipeline that solves the following\n",
        "problem in comparative analysis of gene cluster characterization profiles:\n",
        "\n",
        "Given two COG-spelled sets of genomes S1 and S2, each in a separate file, and also given parameters:\n",
        "\n",
        "**d** â€“ the number of genes in the sought gene cluster\n",
        "\n",
        "**q1** â€“ the required quorum for a gene cluster in genome set S1 (for example, the plasmid genome set).\n",
        "\n",
        "**q2**- the required quorum for a gene cluster in genome set S2 (for example, the chromosome genome set).\n",
        "\n",
        "Objective: Find a gene-cluster GC of cardinality d that appears in at least q1 of the genomes in set S1, and also appears in at least q2 genomes of Set S2, such that the following requirement is obeyed: the characterization profile of gene cluster GC in set S1 differs significantly from the characterization profile of the same gene cluster in set S2."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jnFnTEjDYlv5",
        "outputId": "95f30a95-17b3-4732-e415-0a85aeb68545"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[{'4587', '3694'}, 60.333333333333336]\n"
          ]
        }
      ],
      "source": [
        "import copy\n",
        "\n",
        "def find_GC(s1,s2,d,q1,q2):\n",
        "    global env_info\n",
        "    global taxa_info\n",
        "    global various_gene_orders\n",
        "    gen_uid_dict_s1 = {}\n",
        "    original_data_dict_s1 = {}\n",
        "    cog_dict_s1 = {}\n",
        "    taxa_dict_s1 = {}\n",
        "    habitat_dict_s1 = {}\n",
        "    final_clusters_dict_s1 = {}\n",
        "    all_clusters_dict_s1 = {}\n",
        "    clusters_counter_dict_s1= {}\n",
        "    process_data(s1, gen_uid_dict_s1, original_data_dict_s1)\n",
        "    create_cog_dict(cog_dict_s1)\n",
        "    create_taxa_dict(taxa_dict_s1)\n",
        "    create_habitat_dict(habitat_dict_s1)\n",
        "    find_clusters(original_data_dict_s1, d, q1, final_clusters_dict_s1, all_clusters_dict_s1, clusters_counter_dict_s1) # q=2 is a default value\n",
        "    final_clusters_dict_s1 = {cluster: values for cluster, values in all_clusters_dict_s1.items() if clusters_counter_dict_s1[cluster] >= q1}\n",
        "    clusters_counter_dict_s1 = {cluster: value for cluster, value in clusters_counter_dict_s1.items() if value >= q1}\n",
        "\n",
        "    gen_uid_dict_s2 = {}\n",
        "    original_data_dict_s2 = {}\n",
        "    cog_dict_s2 = {}\n",
        "    taxa_dict_s2 = {}\n",
        "    habitat_dict_s2 = {}\n",
        "    final_clusters_dict_s2 = {}\n",
        "    all_clusters_dict_s2 = {}\n",
        "    clusters_counter_dict_s2= {}\n",
        "    process_data(s2, gen_uid_dict_s2, original_data_dict_s2)\n",
        "    create_cog_dict(cog_dict_s2)\n",
        "    create_taxa_dict(taxa_dict_s2)\n",
        "    create_habitat_dict(habitat_dict_s2)\n",
        "    find_clusters(original_data_dict_s2, d, q2, final_clusters_dict_s2, all_clusters_dict_s2, clusters_counter_dict_s2) # q=2 is a default value\n",
        "    final_clusters_dict_s2 = {cluster: values for cluster, values in all_clusters_dict_s2.items() if clusters_counter_dict_s2[cluster] >= q2}\n",
        "    clusters_counter_dict_s2 = {cluster: value for cluster, value in clusters_counter_dict_s2.items() if value >= q2}\n",
        "    score = 0\n",
        "    GC = {}\n",
        "    for cluster in final_clusters_dict_s1:\n",
        "        score = 0\n",
        "        env_info = {}\n",
        "        taxa_info = {}\n",
        "        various_gene_orders = {}\n",
        "        if cluster in final_clusters_dict_s2:\n",
        "            #print(cluster)\n",
        "            get_gene_orders_dict(cluster, 0, final_clusters_dict_s1, clusters_counter_dict_s1)\n",
        "            get_taxa_info(cluster, 0, final_clusters_dict_s1, taxa_dict_s1, gen_uid_dict_s1)\n",
        "            get_env_info(cluster,0, final_clusters_dict_s1, gen_uid_dict_s1, habitat_dict_s1)\n",
        "            get_gene_orders_dict(cluster, 1, final_clusters_dict_s2, clusters_counter_dict_s2)\n",
        "            get_taxa_info(cluster,1, final_clusters_dict_s2, taxa_dict_s2, gen_uid_dict_s2)\n",
        "            get_env_info(cluster, 1, final_clusters_dict_s2, gen_uid_dict_s2, habitat_dict_s2)\n",
        "            score = calculate_profile_score(cluster, q1,q2)\n",
        "        if score > 60 :\n",
        "            GC = set(cluster)\n",
        "            break\n",
        "    return  [GC, score]\n",
        "\n",
        "def test_GC():\n",
        "    print(find_GC(\"cog_words_plasmid.txt\",\"clean_file.txt\",2,2,2))\n",
        "\n",
        "test_GC()\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TWrI0KvultIm"
      },
      "source": [
        "**4)** Experiment with your program on the data files (i.e., run benchmarks with different\n",
        "values of d, q1, q2 and examine the results) till you find an interesting result"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ncWjqugSmWo6"
      },
      "outputs": [],
      "source": [
        "def experiment_GC_values():\n",
        "    max_score = 0\n",
        "    best_d = 0\n",
        "    best_q1 = 0\n",
        "    best_q2 = 0\n",
        "    for d_val in range(2,6):\n",
        "        for q1_val in range(1,100):\n",
        "            q2_val = q1_val * 10\n",
        "            gen_cluster_list = find_GC(\"cog_words_plasmid.txt\",\"clean_file.txt\",d_val,q1_val,q2_val)\n",
        "            my_score = gen_cluster_list[1]\n",
        "            if my_score > max_score:\n",
        "                max_score = my_score\n",
        "                best_d = d_val\n",
        "                best_q1 = q1_val\n",
        "                best_q2 = q2_val\n",
        "                my_list = gen_cluster_list\n",
        "                print(my_list, best_d, best_q1, best_q2) #test\n",
        "    print(my_list, best_d, best_q1, best_q2)\n",
        "\n",
        "#experiment_GC_values()"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Process the data about the cluster**\n",
        "\n",
        "The cluster: {'1898', '1088', '1209', '1091'}\n",
        "\n",
        "(d: 4, q1: 1, q2: 10)\n",
        "\n",
        "cog information:\n",
        "\n",
        "| COG       | Functional Category         | Role                               | Enzyme                                         |\n",
        "|-----------|-----------------------------|------------------------------------|------------------------------------------------|\n",
        "| COG1898   | Cellular Processes and Signaling | Cell wall/membrane/envelope biogenesis | dTDP-4-dehydrorhamnose 3,5-epimerase or related enzyme |\n",
        "| COG1088   | Cellular Processes and Signaling | Cell wall/membrane/envelope biogenesis | dTDP-D-glucose 4,6-dehydratase                    |\n",
        "| COG1209   | Cellular Processes and Signaling | Cell wall/membrane/envelope biogenesis | dTDP-glucose pyrophosphorylase                    |\n",
        "| COG1091   | Cellular Processes and Signaling | Cell wall/membrane/envelope biogenesis | dTDP-4-dehydrorhamnose reductase                 |\n",
        "\n",
        "\n",
        " These enzymes are interconnected through their involvement in common metabolic pathways and the shared substrates and products in their respective reactions. Let's break down the connections:\n",
        "\n",
        "# Enzyme Information\n",
        "\n",
        "## 1. COG1898: dTDP-4-dehydrorhamnose 3,5-epimerase\n",
        "- **Reaction:** dTDP-4-dehydro-6-deoxy-D-glucose â‡Œ dTDP-4-dehydro-6-deoxy-L-mannose\n",
        "\n",
        "## 2. COG1088: dTDP-glucose 4,6-dehydratase\n",
        "- **Reaction:** dTDP-glucose â‡Œ dTDP-4-dehydro-6-deoxy-D-glucose + H2O\n",
        "- **Connection to COG1898:** COG1088 produces a substrate (dTDP-4-dehydro-6-deoxy-D-glucose) for COG1898.\n",
        "\n",
        "## 3. COG1209: Glucose-1-phosphate thymidylyltransferase\n",
        "- **Reaction:** dTTP + alpha-D-glucose 1-phosphate â‡Œ diphosphate + dTDP-glucose\n",
        "- **Family:** Transferases (nucleotidyltransferases)\n",
        "- **Pathways:** Nucleotide sugars metabolism, streptomycin biosynthesis, polyketide sugar unit biosynthesis.\n",
        "- **Connection to COG1088:** COG1209 produces dTDP-glucose, a substrate for COG1088.\n",
        "\n",
        "## 4. COG1091: dTDP-4-dehydrorhamnose reductase\n",
        "- **Reaction:** dTDP-6-deoxy-L-mannose + NADP+ â‡Œ dTDP-4-dehydro-6-deoxy-L-mannose + NADPH + H+\n",
        "- **Connection to COG1898:** COG1091 shares a product (dTDP-4-dehydro-6-deoxy-L-mannose) with COG1898.\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "Jo3NRSIs0Jlw"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def get_cluster_profile(cluster, q1,q2):\n",
        "    global various_gene_orders, taxa_info, env_info\n",
        "    plasmid_curr = 13\n",
        "    chrom_curr = 17\n",
        "    num_of_orders = 0\n",
        "    index = 0\n",
        "\n",
        "    for order in various_gene_orders:\n",
        "        index = index + 1\n",
        "        print(index,\".order: \",order)\n",
        "        print(\"\")\n",
        "        print(\"p: \", round(various_gene_orders[order][0]),\"%,\",round(various_gene_orders[order][0]/100 *plasmid_curr) )\n",
        "        print(\"c: \", round(various_gene_orders[order][1]),\"%,\", round(various_gene_orders[order][1]/100*chrom_curr))\n",
        "        print(\"\")\n",
        "        data = taxa_info[order]\n",
        "        print(\"phylum\")\n",
        "        for item in data:\n",
        "            my_list = data[item]\n",
        "            print(item)\n",
        "            print(\"p: \", round(my_list[0]*various_gene_orders[order][0]/100),\"%, \",round((my_list[0]*various_gene_orders[order][0]/100)*plasmid_curr/100))\n",
        "            print(\"c: \",round( my_list[1]*various_gene_orders[order][1]/100),\"%, \",round(chrom_curr*(my_list[1]*various_gene_orders[order][1]/100)/100))\n",
        "            print(\"\")\n",
        "        data = env_info[order]\n",
        "        print(\"habitats\")\n",
        "        for item in data:\n",
        "            my_list = data[item]\n",
        "            print(item)\n",
        "            print(\"p: \", round(my_list[0]*various_gene_orders[order][0]/100),\"%,\",round((my_list[0]*various_gene_orders[order][0]/100)*plasmid_curr/100))\n",
        "            print(\"c: \", round(my_list[1]*various_gene_orders[order][1]/100),\"%,\",round(chrom_curr*(my_list[1]*various_gene_orders[order][1]/100)/100))\n",
        "            print(\"\")\n",
        "from tabulate import tabulate\n",
        "\n",
        "\n",
        "def get_cluster_profile_1(cluster, q1, q2):\n",
        "    global various_gene_orders, taxa_info, env_info\n",
        "    plasmid_curr = 13\n",
        "    chrom_curr = 17\n",
        "    num_of_orders = 0\n",
        "    index = 0\n",
        "\n",
        "    table_data = []\n",
        "\n",
        "    for order in various_gene_orders:\n",
        "        index = index + 1\n",
        "        order_data = []\n",
        "        order_data.append(f\"{index}. order: {order}\")\n",
        "\n",
        "        p_percentage = round(various_gene_orders[order][0])\n",
        "        c_percentage = round(various_gene_orders[order][1])\n",
        "\n",
        "        # Check if both plasmid and chromosome percentages are non-zero\n",
        "        if p_percentage != 0 or c_percentage != 0:\n",
        "            order_p_value = round(p_percentage / 100 * plasmid_curr)\n",
        "            order_c_value = round(c_percentage / 100 * chrom_curr)\n",
        "\n",
        "            order_data.append(f\"{p_percentage}%, {order_p_value}\")\n",
        "            order_data.append(f\"{c_percentage}%, {order_c_value}\")\n",
        "            if order_p_value != 0 or order_c_value != 0:\n",
        "                table_data.append(order_data)\n",
        "                table_data.append([\"-----------------------\", \"-----------------------\", \"-----------------------\"])\n",
        "            data = taxa_info[order]\n",
        "            for item in data:\n",
        "                my_list = data[item]\n",
        "                p_phylum = round(my_list[0] * various_gene_orders[order][0] / 100)\n",
        "                c_phylum = round(my_list[1] * various_gene_orders[order][1] / 100)\n",
        "\n",
        "                # Check if both plasmid and chromosome percentages are non-zero\n",
        "                if p_phylum != 0 or c_phylum != 0:\n",
        "                    phylum_p_value = round(p_phylum * plasmid_curr / 100)\n",
        "                    phylum_c_value = round(c_phylum * chrom_curr / 100)\n",
        "                    phylum_data = [item, f\"{p_phylum}%, {phylum_p_value}\", f\"{c_phylum}%, {phylum_c_value}\"]\n",
        "                    if phylum_p_value != 0 or phylum_c_value != 0:\n",
        "                        table_data.append(phylum_data)\n",
        "            table_data.append([\"-----------------------\", \"-----------------------\", \"-----------------------\"])\n",
        "            data = env_info[order]\n",
        "            for item in data:\n",
        "                my_list = data[item]\n",
        "                p_habitat = round(my_list[0] * various_gene_orders[order][0] / 100)\n",
        "                c_habitat = round(my_list[1] * various_gene_orders[order][1] / 100)\n",
        "\n",
        "                # Check if both plasmid and chromosome percentages are non-zero\n",
        "                if p_habitat != 0 or c_habitat != 0:\n",
        "                    habitat_p_value = round(p_habitat * plasmid_curr / 100)\n",
        "                    habitat_c_value = round(c_habitat * chrom_curr / 100)\n",
        "                    habitat_data = [item, f\"{p_habitat}%, {habitat_p_value}\", f\"{c_habitat}%, {habitat_c_value}\"]\n",
        "                    if habitat_p_value != 0 or habitat_c_value != 0:\n",
        "                        table_data.append(habitat_data)\n",
        "            table_data.append([\"________________________\", \"________________________\", \"________________________\"])\n",
        "    headers = [\"Category\", \"Plasmid (%)\", \"Chromosome (%)\"]\n",
        "    print(tabulate(table_data, headers=headers, tablefmt=\"pretty\"))\n",
        "\n",
        "\n",
        "\n",
        "def print_cluster_profile():\n",
        "    global taxa_info\n",
        "    global env_info\n",
        "    global various_gene_orders\n",
        "    various_gene_orders = {}\n",
        "    taxa_info = {}\n",
        "    env_info = {}\n",
        "    cluster_set = frozenset({'1898', '1088', '1209', '1091'})\n",
        "    create_cluster_profile(\"cog_words_plasmid.txt\", cluster_set, taxa_info, env_info, various_gene_orders,0,1)\n",
        "    create_cluster_profile(\"clean_file.txt\", cluster_set, taxa_info, env_info, various_gene_orders,1,10)\n",
        "    get_cluster_profile_1(cluster_set, 1,10)\n",
        "\n",
        "\n",
        "print_cluster_profile()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7NoNtcQg1Crz",
        "outputId": "dcf19190-fb7b-470a-8bd3-3007f9af2a69"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "13\n",
            "159\n",
            "+----------------------------+--------------------------+--------------------------+\n",
            "|          Category          |       Plasmid (%)        |      Chromosome (%)      |\n",
            "+----------------------------+--------------------------+--------------------------+\n",
            "| 1. order: 1088120918981091 |          8%, 1           |          23%, 4          |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|       Proteobacteria       |          8%, 1           |          21%, 4          |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|       Not Annotated        |          8%, 1           |          3%, 1           |\n",
            "|     Marine Environment     |          0%, 0           |          3%, 1           |\n",
            "|     Soil and Sediment      |          0%, 0           |          6%, 1           |\n",
            "|           Plant            |          0%, 0           |          7%, 1           |\n",
            "|  ________________________  | ________________________ | ________________________ |\n",
            "| 2. order: 1898108812091091 |          8%, 1           |          1%, 0           |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|       Acidobacteria        |          8%, 1           |          0%, 0           |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|       Not Annotated        |          8%, 1           |          0%, 0           |\n",
            "|  ________________________  | ________________________ | ________________________ |\n",
            "| 3. order: 1898108810911209 |          46%, 6          |          14%, 2          |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|       Proteobacteria       |          46%, 6          |          14%, 2          |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|           Plant            |          8%, 1           |          5%, 1           |\n",
            "|       Not Annotated        |          15%, 2          |          1%, 0           |\n",
            "|     Soil and Sediment      |          8%, 1           |          4%, 1           |\n",
            "|           Animal           |          8%, 1           |          0%, 0           |\n",
            "|     Marine Environment     |          8%, 1           |          1%, 0           |\n",
            "|  ________________________  | ________________________ | ________________________ |\n",
            "| 4. order: 1898109112091088 |          8%, 1           |          2%, 0           |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|        Spirochaetes        |          8%, 1           |          0%, 0           |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|           Animal           |          8%, 1           |          0%, 0           |\n",
            "|  ________________________  | ________________________ | ________________________ |\n",
            "| 5. order: 1209189810911088 |          8%, 1           |          6%, 1           |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|        Fusobacteria        |          8%, 1           |          0%, 0           |\n",
            "|       Proteobacteria       |          0%, 0           |          4%, 1           |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|     Soil and Sediment      |          8%, 1           |          3%, 1           |\n",
            "|  ________________________  | ________________________ | ________________________ |\n",
            "| 6. order: 1088109118981209 |          8%, 1           |          0%, 0           |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|         Firmicutes         |          8%, 1           |          0%, 0           |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|            Host            |          8%, 1           |          0%, 0           |\n",
            "|  ________________________  | ________________________ | ________________________ |\n",
            "| 7. order: 1209189810881091 |          15%, 2          |          22%, 4          |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|       Proteobacteria       |          15%, 2          |          3%, 1           |\n",
            "|         Firmicutes         |          0%, 0           |          19%, 3          |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|     Soil and Sediment      |          8%, 1           |          3%, 1           |\n",
            "|            Host            |          8%, 1           |          7%, 1           |\n",
            "|           Human            |          0%, 0           |          3%, 1           |\n",
            "|           Animal           |          0%, 0           |          4%, 1           |\n",
            "|       Not Annotated        |          0%, 0           |          3%, 1           |\n",
            "|  ________________________  | ________________________ | ________________________ |\n",
            "| 8. order: 1088109112091898 |          0%, 0           |          23%, 4          |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|       Proteobacteria       |          0%, 0           |          23%, 4          |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|     Soil and Sediment      |          0%, 0           |          5%, 1           |\n",
            "|       Not Annotated        |          0%, 0           |          3%, 1           |\n",
            "|            Host            |          0%, 0           |          6%, 1           |\n",
            "|        Fresh water         |          0%, 0           |          3%, 1           |\n",
            "|  ________________________  | ________________________ | ________________________ |\n",
            "| 9. order: 1088120910911898 |          0%, 0           |          6%, 1           |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|       Proteobacteria       |          0%, 0           |          6%, 1           |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|  ________________________  | ________________________ | ________________________ |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|  ________________________  | ________________________ | ________________________ |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|  ________________________  | ________________________ | ________________________ |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|  ________________________  | ________________________ | ________________________ |\n",
            "|  -----------------------   | -----------------------  | -----------------------  |\n",
            "|  ________________________  | ________________________ | ________________________ |\n",
            "+----------------------------+--------------------------+--------------------------+\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "The genes RfbD, RfbB, RfbC, and RmlA1, identified in the study : **O-Antigen-Dependent Colicin Insensitivity of Uropathogenic Escherichia coli.**\n",
        "These play roles in the biosynthesis of lipopolysaccharide (LPS), a crucial component of the outer membrane of Gram-negative bacteria. These genes are involved in the O-antigen synthesis pathway.\n",
        "\n",
        "The O-antigen is the outermost component of the LPS molecule and contributes to the structural and functional properties of the bacterial outer membrane. The study suggests that the density of O-antigen is a critical factor in determining the sensitivity of uropathogenic Escherichia coli (UPEC) to colicins.\n",
        "\n",
        "The density of the O-antigen, influenced by the activities of these genes, affects the ability of colicins to bind to their receptors and, consequently, the sensitivity of the bacteria to colicin-mediated killing.\n",
        "\n",
        "After observation of the cluster profile we found, considering the different orders, habitats, and taxa information, we can notice some findings.\n",
        "\n",
        "There are some data we found only in the plasmid:\n",
        "\n",
        "â€¢\t(order number 2 ,1,   Acidobacteria, Not Annotated)\n",
        "â€¢\t(order number 3 ,1, Proteobacteria, Animal)\n",
        "â€¢\t(order number 4, 1,   Spirochaetes, Animal)\n",
        "â€¢\t(order number 6,1, Firmicutes, Host)\n",
        "\n",
        "We noticed different distributions of the factors we tested, and it also appears that the gene cluster is more abundant in plasmid relative to chromosomes. Therefore, we suggest several hypotheses of advantages that clusters give in relation to the article and the role of the genes in the biosynthesis of lipopolysaccharides (LPS):\n",
        "1.\tFunctionality in Variable Environments: The prevalence of LPS biosynthesis genes on plasmids, observed across various habitats (e.g. Acidobacteria, Proteobacteria, Spirochaetes, and Firmicutes), suggests an adaptive advantage for bacteria in fluctuating environments. Plasmids enable rapid modulation of cell surface structures, crucial for survival in diverse conditions.\n",
        "2.\tHorizontal Gene Transfer (HGT): The consistent presence of these genes on plasmids indicates their potential for horizontal gene transfer (HGT) among bacterial populations. This genetic exchange allows for the acquisition of traits related to outer membrane composition and colicin insensitivity across different taxa.\n",
        "3.\tEcological Niche Adaptation: The data showing plasmid localization of LPS biosynthesis genes in taxa associated with specific ecological niches (e.g. Animal ) suggests their role in niche adaptation. This localization likely enhances bacterial fitness within these niches, possibly through evasion of host immune responses.\n",
        "4.\tSelective Pressure and Fitness Cost: Despite the fitness cost associated with plasmid maintenance, the widespread distribution of these genes on plasmids indicates that the benefits outweigh the costs. The selective pressures imposed by diverse environments likely drive the retention of these genes on mobile genetic elements.\n",
        "\n",
        "In summary, the abundance of the gene cluster {1898, 1088, 1209, 1091} in plasmids compared to chromosomes suggests that these genes provide selective advantages to bacteria when located on plasmids.\n",
        "_________________________________________\n",
        "\n"
      ],
      "metadata": {
        "id": "HVpRXndbgkD7"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "- Related article: **O-Antigen-Dependent Colicin Insensitivity of Uropathogenic Escherichia coli**\n",
        "\n",
        "**Citation**: Sharp C, Boinett C, Cain AHousden NG, Kumar S, Turner K, Parkhill J, Kleanthous C.2019.O-Antigen-Dependent Colicin Insensitivity of Uropathogenic Escherichia coli. J Bacteriol201:10.1128/jb.00545-18.https://doi.org/10.1128/jb.00545-18\n",
        "\n",
        "**Summary**:\n",
        "The article discusses the phenomenon of colicin insensitivity in Escherichia coli (E. coli) strains, particularly focusing on uropathogenic E. coli (UPEC) sequence type 131 (ST131). Colicins are antimicrobial proteins secreted by E. coli, and they have the potential to be used as narrow-spectrum antimicrobials.\n",
        "\n",
        "The researchers demonstrate that this insensitivity can be overcome by making minor changes to the environmental conditions.\n",
        "\n",
        "They found that the production of O-antigen is a key factor in colicin insensitivity but that slight modifications to growth conditions can render the organism sensitive to colicins. The reintroduction of O-antigen into E. coli K-12 demonstrated that the density of O-antigen is a dominant factor governing colicin insensitivity.\n",
        "\n",
        "One of the measures that was used in the research: A transposon-directed insertion sequencing (TraDIS) approach was employed to identify genes associated with colicin insensitivity in UPEC ST131. Thirty-three genes (four of them are the genes in the cluster we found) were mapped to operons responsible for lipopolysaccharide (LPS) biosynthesis, O-antigen biosynthesis, and other outer membrane-related processes.\n",
        "\n",
        "Fluorescence microscopy revealed that the presence of O-antigen diminished colicin binding to the outer membrane receptor (BtuB). Changes in O-antigen density influenced colicin binding and sensitivity.\n",
        "\n",
        "In summary, the study provides insights into the mechanisms and environmental factors influencing colicin insensitivity in UPEC ST131, shedding light on the complex interplay between bacterial genetics, outer membrane composition, and environmental conditions."
      ],
      "metadata": {
        "id": "9AhqtdPhv5l3"
      }
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}